File: Server/ConcordiaCurriculumManager/Models/Curriculum/CourseGroupings/CourseGrouping.cs
using ConcordiaCurriculumManager.DTO.CourseGrouping;
using ConcordiaCurriculumManager.Filters.Exceptions;
using ConcordiaCurriculumManager.Models.Curriculum.Dossiers;
using NpgsqlTypes;

namespace ConcordiaCurriculumManager.Models.Curriculum.CourseGroupings;

public class CourseGrouping : BaseModel
{
    public required Guid CommonIdentifier { get; set; }
    public required string Name { get; set; }
    public required string RequiredCredits { get; set; }
    public required bool IsTopLevel { get; set; }
    public required SchoolEnum School { get; set; }
    public string? Description { get; set; }
    public string? Notes { get; set; }
    public required CourseGroupingStateEnum State { get; set; }
    public int? Version { get; set; }
    public required bool Published { get; set; }
    public required ICollection<CourseGroupingReference> SubGroupingReferences { get; set; } = new List<CourseGroupingReference>();
    public required ICollection<CourseGrouping> SubGroupings { get; set; } = new List<CourseGrouping>();
    public required ICollection<CourseIdentifier> CourseIdentifiers { get; set; } = new List<CourseIdentifier>();
    public required ICollection<Course> Courses { get; set; } = new List<Course>();
    public CourseGroupingRequest? CourseGroupingRequest { get; set; }

    public static CourseGrouping CreateCourseGroupingFromCreationDTO(Guid id, CourseGroupingInputDTO dto) =>
        CreateCourseGroupingFromDTO(id, dto, Guid.NewGuid(), CourseGroupingStateEnum.NewCourseGroupingProposal);

    public static CourseGrouping CreateCourseGroupingModificationFromModificationDTO(Guid id, CourseGroupingModificationInputDTO dto) =>
        CreateCourseGroupingFromDTO(id, dto, dto.CommonIdentifier, CourseGroupingStateEnum.CourseGroupingChangeProposal);

    public static CourseGrouping CreateCourseGroupingDeletionFromModificationDTO(Guid id, CourseGroupingModificationInputDTO dto) =>
        CreateCourseGroupingFromDTO(id, dto, dto.CommonIdentifier, CourseGroupingStateEnum.CourseGroupingDeletionProposal);

    private static CourseGrouping CreateCourseGroupingFromDTO(Guid id,CourseGroupingInputDTO dto, Guid commonId, CourseGroupingStateEnum state)
    {
        return new CourseGrouping
        {
            Id = id,
            CommonIdentifier = commonId,
            Name = dto.Name,
            RequiredCredits = dto.RequiredCredits,
            IsTopLevel = dto.IsTopLevel,
            School = dto.School,
            Description = dto.Description,
            Notes = dto.Notes,
            State = state,
            Version = null,
            Published = false,
            SubGroupingReferences = CourseGroupingReference.CreateCourseGroupingReferencesFromDTO(id, dto.SubGroupingReferences),
            SubGroupings = new List<CourseGrouping>(),
            CourseIdentifiers = CourseIdentifier.CreateCourseIdentifiersFromDTO(dto.CourseIdentifiers),
            Courses = new List<Course>()
        };
    }

    public bool IsCourseGroupingStateFinalized() => State == CourseGroupingStateEnum.Accepted || State == CourseGroupingStateEnum.Deleted;

    public void MarkAsPublished()
    {
        Published = true;

        VerifyCourseGroupingIsValidOrThrow();
    }

    public void MarkAsUnpublished()
    {
        Published = false;
    }

    private void VerifyCourseGroupingIsValidOrThrow()
    {
        if (Published)
        {
            if (IsCourseGroupingStateFinalized() && Version != null)
            {
                foreach (var courseGrouping in SubGroupings)
                {
                    if (!courseGrouping.Published)
                    {
                        throw new ArgumentException($"Cannot publish a course grouping with unpublished sub-course grouping");
                    }
                }

                foreach (var course in Courses)
                {
                    if (!course.Published)
                    {
                        throw new ArgumentException($"Cannot publish a course grouping with unpublished courses");
                    }
                }

                return;
            }

            throw new ArgumentException("The course group is published but does not have an appropriate course state or version");
        }
    }
}

public enum SchoolEnum
{
    [PgName(nameof(GinaCody))]
    GinaCody,

    [PgName(nameof(ArtsAndScience))]
    ArtsAndScience,

    [PgName(nameof(FineArts))]
    FineArts,

    [PgName(nameof(JMSB))]
    JMSB
}

public enum CourseGroupingStateEnum
{
    [PgName(nameof(Accepted))]
    Accepted,

    [PgName(nameof(NewCourseGroupingProposal))]
    NewCourseGroupingProposal,

    [PgName(nameof(CourseGroupingChangeProposal))]
    CourseGroupingChangeProposal,

    [PgName(nameof(CourseGroupingDeletionProposal))]
    CourseGroupingDeletionProposal,

    [PgName(nameof(Deleted))]
    Deleted
}

File: App/src/models/courseGrouping.ts
import { Course } from "./course";

export enum SchoolEnum {
    GinaCody = 0,
    ArtsAndScience = 1,
    FineArts = 2,
    JMSB = 3,
}
export const SchoolEnumArray = [
    { value: SchoolEnum.GinaCody, label: "Gina Cody" },
    { value: SchoolEnum.ArtsAndScience, label: "Arts and Science" },
    { value: SchoolEnum.FineArts, label: "Fine Arts" },
    { value: SchoolEnum.JMSB, label: "JMSB" },
];

export enum CourseGroupingStateEnum {
    Accepted = 0,
    NewCourseGroupingProposal = 1,
    CourseGroupingChangeProposal = 2,
    CourseGroupingDeletionProposal = 3,
    Deleted = 4,
}

export enum GroupingTypeEnum {
    subGrouping = 0,
    optionalGrouping = 1,
}

export interface CourseIdentifierDTO {
    subject?: string | null;
    catalog?: number | null;
    concordiaCourseId: number;
}

export interface CourseGroupingReferenceDTO {
    id?: string;
    parentGroupId?: string;
    name?: string | null;
    childGroupCommonIdentifier: string;
    groupingType: GroupingTypeEnum;
}

export interface CourseGroupingReferenceInputDTO {
    childGroupCommonIdentifier: string;
    groupingType: GroupingTypeEnum;
}

export interface CourseGroupingDTO {
    id: string;
    commonIdentifier: string;
    name: string;
    requiredCredits: string;
    isTopLevel: boolean;
    school: SchoolEnum;
    description: string | null;
    notes: string | null;
    state: CourseGroupingStateEnum;
    version: number | null;
    published: boolean;
    subGroupingReferences: CourseGroupingReferenceDTO[];
    subGroupings: CourseGroupingDTO[];
    courseIdentifiers: CourseIdentifierDTO[];
    courses: Course[];
    createdDate: Date;
    modifiedDate: string;
}

export interface CourseGroupingRequestDTO {
    id: string;
    dossierId: string;
    rationale: string;
    resourceImplication: string;
    comment: string;
    conflict: string;
    courseGrouping: CourseGroupingDTO;
    requestType?: number;
}

export interface CourseGroupingModificationInputDTO {
    name: string;
    requiredCredits: string;
    isTopLevel: boolean;
    school: SchoolEnum;
    description: string | null;
    notes: string | null;
    subGroupingReferences: CourseGroupingReferenceInputDTO[];
    courseIdentifiers: CourseIdentifierDTO[];
    commonIdentifier: string;
}

export interface CourseGroupingInputDTO {
    name: string | null;
    requiredCredits: string | null;
    isTopLevel: boolean;
    school: SchoolEnum;
    description: string | null;
    notes: string | null;
    subGroupingReferences: CourseGroupingReferenceInputDTO[];
    courseIdentifiers: CourseIdentifierDTO[];
}

export interface CourseGroupingCreationRequestDTO {
    dossierId: string;
    rationale: string | null;
    resourceImplication: string | null;
    comment: string | null;
    courseGrouping: CourseGroupingInputDTO;
}

export interface CourseGroupingModificationRequestDTO {
    dossierId: string;
    rationale: string | null;
    resourceImplication: string | null;
    comment: string | null;
    courseGrouping: CourseGroupingModificationInputDTO;
}

export interface MultiCourseGroupingDTO {
    data: CourseGroupingDTO[];
}
